deno2/deno.gni

```diff
@@ -0,0 +1,197 @@
+    forward_variables_from(invoker, "*")
+    script = "js/run_node.py"
+  }
+}
+
+# Template to generate different V8 snapshots based on different runtime flags.
+# Can be invoked with run_mksnapshot(<name>). The target will resolve to
+# run_mksnapshot_<name>. If <name> is "default", no file suffixes will be used.
+# Otherwise files are suffixed, e.g. embedded_<name>.cc and
+# snapshot_blob_<name>.bin.
+#
+# The template exposes the variables:
+#   args: additional flags for mksnapshots
+#   embedded_suffix: a camel case suffix for method names in the embedded
+#       snapshot.
+template("create_snapshot") {
+  name = target_name
+  suffix = "_$name"
+  action("create_snapshot_" + name) {
+    forward_variables_from(invoker,
+                           [
+                             "testonly",
+                             "deps",
+                           ])
+    visibility = [ ":*" ]  # Only targets in this file can depend on this.
+    deps += [ ":snapshot_creator" ]
+    script = "v8/tools/run.py"
+    data = []
+    exe = rebase_path(get_label_info(":snapshot_creator", "root_out_dir") +
+                      "/snapshot_creator")
+    natives_in_bin = "$root_out_dir/natives_blob.bin"
+    snapshot_in_bin = "$root_out_dir/snapshot_blob.bin"
+    natives_out_cc = "$target_gen_dir/natives${suffix}.cc"
+    snapshot_out_cc = "$target_gen_dir/snapshot${suffix}.cc"
+    sources = [
+      invoker.js,
+    ]
+    outputs = [
+      natives_out_cc,
+      snapshot_out_cc,
+    ]
+    args = [
+      exe,
+      rebase_path(invoker.js, root_build_dir),
+      rebase_path(natives_in_bin, root_build_dir),
+      rebase_path(snapshot_in_bin, root_build_dir),
+      rebase_path(natives_out_cc, root_build_dir),
+      rebase_path(snapshot_out_cc, root_build_dir),
+    ]
+
+    # To debug snapshotting problems:
+    #  args += ["--trace-serializer"]
+    data = [
+      invoker.js,
+    ]
+  }
+}
+
+template("rust_crate") {
+  crate_type = invoker.crate_type
+  source_root = invoker.source_root
+  action(target_name) {
+    script = "v8/tools/run.py"
+    depfile = "$target_gen_dir/$target_name.d"
+    sources = [
+      source_root,
+    ]
+    outputs = []
+
+    args = [
+      "rustc",
+      rebase_path(source_root, root_build_dir),
+      "--crate-name=$target_name",
+      "--crate-type=$crate_type",
+      "--emit=dep-info=" + rebase_path(depfile, root_build_dir),
+    ]
+
+    # We only use staticlib for the special "empty" lib.
+    if (crate_type == "staticlib") {
+      staticlib = "$target_out_dir/$target_name.a"
+      outputs += [ staticlib ]
+      args += [ "--emit=link=" + rebase_path(staticlib, root_build_dir) ]
+    }
+
+    if (crate_type == "lib" || crate_type == "bin") {
+      obj = "$target_out_dir/$target_name.o"
+      outputs += [ obj ]
+      args += [ "--emit=obj=" + rebase_path(obj, root_build_dir) ]
+    }
+
+    if (crate_type == "lib") {
+      rlib = "$target_out_dir/$target_name.rlib"
+      outputs += [ rlib ]
```

Comment:
Below where we do `rust_objs += get_target_outputs(":" + bin_target)` 

won't this cause the rlib to get passed into the linker?

Reply:
I think it just tells gn/ninja it's an output. My linker invocation:

```
../../third_party/llvm-build/Release+Asserts/bin/lld-link.exe /nologo /OUT:./deno_rs.exe /PDB:./deno_rs.exe.pdb obj/libdeno.lib advapi32.lib comdlg32.lib dbghelp.lib dnsapi.lib gdi32.lib msimg32.lib odbc32.lib odbccp32.lib oleaut32.lib psapi.lib shell32.lib shlwapi.lib user32.lib usp10.lib uuid.lib version.lib wininet.lib winmm.lib winspool.lib ws2_32.lib delayimp.lib kernel32.lib ole32.lib  obj/deno_rs_stdlib.a obj/deno_rs_bin.o obj/libc.o "userenv.lib gg" --color-diagnostics /WX /DEBUG /MACHINE:X64 /fastfail /FIXED:NO /ignore:4199 /ignore:4221 /NXCOMPAT /DYNAMICBASE /INCREMENTAL /OPT:NOREF /SUBSYSTEM:CONSOLE,5.02 /LIBPATH:../../third_party/llvm-build/Release+Asserts/lib/clang/7.0.0/lib/windows /LIBPATH:"C:/Program Files (x86)/Windows Kits/10/lib/10.0.17134.0/um/x64" /LIBPATH:"C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.14.26428/lib/x64" /LIBPATH:"C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.14.26428/ATLMFC/lib/x64"
```

Reply:
ok let's leave it there.